R/fill.R

```diff
@@ -82,22 +83,39 @@
 #'   3, "Danielle",    "Observer",   NA
 #' )
 #'
-#' # The values are inconsistently missing by position within the group
-#' # Use .direction = "downup" to fill missing values in both directions
+#' # The values are inconsistently missing by position within the `group`.
+#' # Use `.direction = "downup"` to fill missing values in both directions
+#' # and `.by = group` to apply the fill per group.
+#' squirrels %>%
+#'   fill(n_squirrels, .direction = "downup", .by = group)
+#'
+#' # If you want, you can also supply a data frame grouped with `group_by()`,
+#' # but don't forget to `ungroup()`!
 #' squirrels %>%
 #'   dplyr::group_by(group) %>%
 #'   fill(n_squirrels, .direction = "downup") %>%
 #'   dplyr::ungroup()
-#'
-#' # Using `.direction = "updown"` accomplishes the same goal in this example
-fill <- function(data, ..., .direction = c("down", "up", "downup", "updown")) {
-  check_dots_unnamed()
+fill <- function(data,
+                 ...,
+                 .direction = c("down", "up", "downup", "updown")) {
   UseMethod("fill")
 }
 
+#' @inheritParams dplyr::mutate
+#' @rdname fill
 #' @export
-fill.data.frame <- function(data, ..., .direction = c("down", "up", "downup", "updown")) {
-  vars <- tidyselect::eval_select(expr(c(...)), data, allow_rename = FALSE)
+fill.data.frame <- function(data,
+                            ...,
+                            .direction = c("down", "up", "downup", "updown"),
+                            .by = NULL) {
```

Comment:
Putting `.by` in the `.data.frame` method rather than the generic to try and be as least disruptive as possible for S3 method authors like dbplyr

Reply:
TODO: We did add `.by` to the `nest()` generic, so maybe we should look at this a little more.

Maybe it isn't as disruptive as we thought due to the presence of the `...` in the generic.

Reply:
According to my old research on this, it doesn't look like we will cause a warning for other S3 method authors if we add `.by` to the generic:

https://github.com/DavisVaughan/methodtest/blob/0010788d798e9ff851a871db3cd35e8ee1cb4dd6/R/methods.R#L55-L67

They will just have to opt in to support of `.by` on their own over time or catch `.by` and error if its not `NULL` if they don't support it.

Reply:
Ok, I've gone back to `.by` in the generic of `fill()` based on the above analysis. That means that for dbplyr we will see something like this where `.by` is captured in the `...` until dbplyr updates its S3 method to handle `.by` (possibly by just erroring).

By some weird combination of events this passes through dbplyr without error, I've opened https://github.com/tidyverse/dbplyr/issues/1536

```r
df <- data.frame(x = c(1, 2), y = c(2, 2), z = c(1, NA))
df_sqlite <- tbl_lazy(df, con = simulate_sqlite())
df_sqlite |> window_order(x) |> fill(z, .by = y)

<SQL>
SELECT `x`, `y`, MAX(`z`) OVER (PARTITION BY `..dbplyr_partition_1`) AS `z`
FROM (
  SELECT
    `df`.*,
    SUM(CASE WHEN ((`z` IS NULL)) THEN 0 ELSE 1 END) OVER (ORDER BY `x` ROWS UNBOUNDED PRECEDING) AS `..dbplyr_partition_1`,
    SUM(CASE WHEN ((`.by` IS NULL)) THEN 0 ELSE 1 END) OVER (ORDER BY `x` ROWS UNBOUNDED PRECEDING) AS `..dbplyr_partition_2`
  FROM `df`
) AS `q01`
```